thorlcr/slave/slavmain.cpp

```diff
@@ -1049,116 +1008,107 @@ class CKJService : public CSimpleInterfaceOf<IKJService>, implements IThreaded,
     CKeyLookupContext *ensureKeyLookupContext(CJobBase &job, const CLookupKey &key, MemoryBuffer &createCtxMb, bool *created=nullptr)
     {
         CriticalBlock b(lCCrit);
-        CKeyLookupContext *keyLookupContext = keyLookupContextsHT.find(key);
-        if (keyLookupContext)
+        auto it = keyLookupContextsHT.find(key);
+        if (it != keyLookupContextsHT.end())
         {
             if (created)
                 *created = false;
-            return LINK(keyLookupContext);
+            return LINK(it->second);
         }
         if (created)
             *created = true;
-        activity_id id = key.id; // JCSMORE annoying, needed because IMPLEMENT_SUPERHASHTABLEOF_REF_FIND doesn't define const (should look at)
-        Linked<CActivityContext> activityCtx = activityContextsHT.find(id);
-        if (!activityCtx)
-        {
-            activityCtx.setown(createActivityContext(job, key.id, createCtxMb));
-            activityContextsHT.replace(*activityCtx); // NB: does not link/take ownership
-        }
-        keyLookupContext = createLookupContext(activityCtx, key);
-        keyLookupContextsHT.replace(*keyLookupContext); // NB: does not link/take ownership
+        Owned<CActivityContext> activityCtx = ensureActivityContext(job, key.id, createCtxMb);
+        CKeyLookupContext *keyLookupContext = createLookupContext(activityCtx, key);
+        keyLookupContextsHT.insert({key, keyLookupContext}); // NB: does not link/take ownership
         return keyLookupContext;
     }
     void removeActivityContext(CActivityContext *activityContext)
     {
         CriticalBlock b(lCCrit);
-        activityContextsHT.removeExact(activityContext);
+        activityContextsHT.erase(activityContext->queryId());
     }
     void freeActivityContext(CActivityContext *_activityContext)
     {
         Owned<CActivityContext> activityContext = _activityContext;
         if (!activityContext->IsShared())
```

Comment:
I think this test needs to be inside the critical block - otherwise another thread could hit it in the cache before it is removed.

Reply:
agree, will change.